improver/regrid/idw.py

```diff
@@ -0,0 +1,219 @@
+    )
+
+    # There may be some output points with no matching nearby surface type (lakes/islands)
+    # Output an array of these so that the calling function can apply further processing to those
+    count_matching_surface = np.count_nonzero(surface_type_mask_updates, axis=1)
+    points_with_no_match = (np.where(count_matching_surface == 0))[0]
+    output_points_no_match = idw_out_indexes[points_with_no_match]
+
+    # Apply inverse distance weighting to points that do have matching surface type input
+    points_with_match = (np.where(count_matching_surface > 0))[0]
+    output_points_match = idw_out_indexes[points_with_match]
+
+    # Convert mask to be true where input points should not be considered
+    not_mask = np.logical_not(surface_type_mask_updates[points_with_match])
+
+    # Replace distances with infinity where they should not be used
+    masked_distances = np.where(
+        not_mask, np.float32(np.inf), distances_updates[points_with_match]
+    )
+
+    # Add a small amount to all distances to avoid division by zero when taking the inverse
+    masked_distances += np.finfo(np.float32).eps
+    # Invert the distances, sum the k surrounding points, scale to produce weights
+    inv_distances = 1.0 / masked_distances
+    # add power 1.80 for inverse diatance weight
+    optimum_power = 1.80
+    inv_distances_power = np.power(inv_distances, optimum_power)
+    inv_distances_sum = np.sum(inv_distances_power, axis=1)
+    inv_distances_sum = 1.0 / inv_distances_sum
+    weights_idw = inv_distances_power * inv_distances_sum.reshape(-1, 1)
+
+    # Update indexes and weights with new values
+    indexes[output_points_match] = indexes_updates[points_with_match]
+    weights[output_points_match] = weights_idw
+    return indexes, weights, output_points_no_match
+
+
+def nearest_input_pts(in_latlons, out_latlons, k):
+    """
+    Find k nearest source (input) points to each target (output)
+    point, using a KDtree
+
+    Args:
+        in_latlons (numpy.ndarray):
+            Source grid points' latitude-longitudes (N x 2)
+        out_latlons (numpy.ndarray):
+            Target grid points' latitude-longitudes (M x 2)
+        k (int):
+            Number of points surrounding each output point
+
+    Return:
+        Tuple[numpy.ndarray, numpy.ndarray]:
+            Distances from target grid point to source grid points and indexes
+            of those points (M x K)
+    """
+    # Convert input latitude and longitude to XYZ coordinates, then create KDtree
+    in_x, in_y, in_z = ecef_coords(in_latlons[:, 0].flat, in_latlons[:, 1].flat)
+    in_coords = np.c_[in_x, in_y, in_z]
+    in_kdtree = KDTree(in_coords)
+    # Convert output to XYZ and query the KDtree for nearby input points
+    out_x, out_y, out_z = ecef_coords(out_latlons[:, 0].flat, out_latlons[:, 1].flat)
+    out_coords = np.c_[out_x, out_y, out_z]
+    distances, indexes = in_kdtree.query(out_coords, k)
```

Comment:
IMPROVER code is generally set up so that each task corresponds to a single CPU thread and parallelism is handled at the suite level.
I'd suggest avoiding multiple workers if possible for simplicity.
If there's an identified performance bottleneck here, the `workers` option could be handy, but doing that would need special handling for this task in the suite.